bundles/automation/org.eclipse.smarthome.automation.providers/src/main/java/org/eclipse/smarthome/automation/core/provider/RuleResourceBundleImporter.java

```diff
@@ -0,0 +1,160 @@
+    }
+
+    @Override
+    public void close() {
+        if (adminTracker != null) {
+            adminTracker.close();
+            adminTracker = null;
+        }
+        super.close();
+    }
+
+    /**
+     * @see org.eclipse.smarthome.automation.core.provider.GeneralResourceBundleProvider#addingService(ServiceReference)
+     */
+    @Override
+    public Object addingService(ServiceReference reference) {
+        Object service = bc.getService(reference);
+        if (service instanceof RuleRegistry) {
+            ruleRegistry = (RuleRegistry) service;
+            return service;
+        }
+        if (reference.getProperty(Parser.PARSER_TYPE).equals(Parser.PARSER_RULE)) {
+            return super.addingService(reference);
+        }
+        return null;
+    }
+
+    /**
+     * @see org.osgi.util.tracker.ServiceTrackerCustomizer#removedService(ServiceReference, Object)
+     */
+    @Override
+    public void removedService(ServiceReference reference, Object service) {
+        if (service instanceof RuleRegistry) {
+            ruleRegistry = null;
+        } else {
+            super.removedService(reference, service);
+        }
+    }
+
+    @Override
+    protected void processAutomationProvider(Bundle bundle) {
+        String parserType = (String) bundle.getHeaders().get(
+                AutomationResourceBundlesEventQueue.AUTOMATION_RESOURCES_HEADER);
+        Parser parser = parsers.get(parserType);
+        if (parser == null || ruleRegistry == null) {
+            synchronized (waitingProviders) {
+                waitingProviders.put(new Long(bundle.getBundleId()), bundle);
+            }
+            return;
+        }
+        Enumeration<URL> urls = bundle.findEntries(path, null, false);
+        if (urls == null) {
+            return;
+        }
+        while (urls.hasMoreElements()) {
+            URL url = urls.nextElement();
+            try {
+                importData(parser, new InputStreamReader(url.openStream()));
+            } catch (IOException e) {
+                log.error("Can't read from URL " + url, e);
+            }
+        }
+    }
+
+    /**
+     * @see org.eclipse.smarthome.automation.core.provider.GeneralResourceBundleProvider#getUID(java.lang.Object)
+     */
+    @Override
+    protected String getUID(Object providedObject) {
+        return ((Rule) providedObject).getUID();
+    }
+
+    /**
+     *
+     * @param parser
+     * @param inputStreamReader
+     */
+    private Set<Status> importData(Parser parser, InputStreamReader inputStreamReader) {
+        Set<Status> providedRulesStatus = parser.importData(inputStreamReader);
+        if (providedRulesStatus != null && !providedRulesStatus.isEmpty()) {
+            Iterator i = providedRulesStatus.iterator();
+            while (i.hasNext()) {
+                Status status = (Status) i.next();
+                Rule rule = (Rule) status.getResult();
+                if (rule != null)
+                    ruleRegistry.add(rule);
```

Comment:
If the ruleRegistry has all rules persisted in its database and the system is restarted, you will see an exception here, because the rule already exists in the registry (as it is being loaded by the ManagedRuleProvider at startup).

I am actually a bit undecided about what is the best way to handle this. Should the rule in the registry be deleted? But this might be the one that the user has already customized. So probably the rule should be kept as it is and the rule from this bundle should only be added if no rule with the same id already exists.